include/pmacc/mpi/GatherSlice.hpp

```diff
@@ -0,0 +1,260 @@
+                MPI_CHECK(MPI_Group_free(&group));
+                MPI_CHECK(MPI_Group_free(&newgroup));
+
+                return this->isMaster();
+            }
+
+
+            /** gather data
+             *
+             * Must be called by all participating MPI ranks.
+             * If a non-participating MPI rank is calling the method the returned buffer will be empty.
+             * @attention The master rank will allocate memory host memory for the received data.
+             *
+             * @tparam T_DataType Slice buffer data type.
+             * @param localInputSlice Buffer with local slice data. Buffer memory must be contiguous without line
+             * paddings. Buffer extents can be different for each MPI rank.
+             * @param globalSliceExtent extent in elements of the global slice
+             * @param localSliceOffset local offset in elements relative to the global slice origin
+             * @return shared pointer to host buffer with gathered slice data (only master has valid data)
+             */
+            template<typename T_DataType>
+            auto gatherSlice(
+                HostBuffer<T_DataType, DIM2>& localInputSlice,
+                DataSpace<DIM2> globalSliceExtent,
+                DataSpace<DIM2> localSliceOffset) const
+            {
+                using ValueType = T_DataType;
+                // guard against wrong usage, only ranks which are participating into the gather are allowed
+                if(!isParticipating())
+                    return std::shared_ptr<HostBuffer<ValueType, DIM2>>{};
+
+                // avoid deadlock between not finished pmacc tasks and mpi blocking collectives
+                eventSystem::getTransactionEvent().waitForFinished();
+                // get number of elements per participating mpi rank
+                auto extentPerDevice = std::vector<DataSpace<DIM2>>(numRanksInPlane);
+
+                auto localSliceSize = localInputSlice.getDataSpace();
+
+                // gather extents
+                MPI_CHECK(MPI_Gather(
+                    reinterpret_cast<int*>(&localSliceSize),
+                    2,
+                    MPI_INT,
+                    reinterpret_cast<int*>(extentPerDevice.data()),
+                    2,
+                    MPI_INT,
+                    0,
+                    gatherComm));
+
+                auto offsetPerDevice = std::vector<DataSpace<DIM2>>(numRanksInPlane);
+
+                // gather offsets
+                MPI_CHECK(MPI_Gather(
+                    reinterpret_cast<int*>(&localSliceOffset),
+                    2,
+                    MPI_INT,
+                    reinterpret_cast<int*>(offsetPerDevice.data()),
+                    2,
+                    MPI_INT,
+                    0,
+                    gatherComm));
+
+                std::vector<int> displs(numRanksInPlane);
+                std::vector<int> count(numRanksInPlane);
+                // @todo replace by std::scan
```

Comment:
Is this a todo for the future or to be done before merging?

Reply:
it is for the future